pipeline/common/datasets.py

```diff
@@ -90,7 +90,12 @@ def shuffle_with_max_lines(
         if len(lines) == max_lines:
             break
 
-    random.shuffle(lines)
+    # random.shuffle requires random access via indexing
+    # deque supports fast adding/removing from its ends with O(1)
+    # but indexing is O(N) which is too slow for shuffling large arrays
+    lines_list = list(lines)
+    random.shuffle(lines_list)
+    lines = deque(lines_list)
```

Comment:
Suggestion (optional): We could `del lines_list` and `del lines` here and other places to make sure the garbage collector could collect any of these lists or deques that we don't need anymore if there is memory pressure.

Reply:
Ok, I added a couple of extra optimizations.